streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1321,11 +1322,22 @@ int process(final int maxNumRecords, final Time time) {
                 totalProcessed += processed;
                 task.recordProcessBatchTime(now - then);
             }
+            successfullyProcessed.add(task);
         }
-
+        successfullyProcessed.clear();
         return totalProcessed;
     }
 
+    public int commitSuccessfullyProcessedTasks() {
+        final int committed = commit(successfullyProcessed);
```

Comment:
@mjsax please double check my understanding here since it's been a while since I thought about/worked with this stuff, but IIRC it's important to note that under the current Producer semantics at least, we can only do this kind of partial commit when using ALOS or EOS-v1. In KIP-447/EOS-v2, if one task/partition is bad then the transaction will need to be aborted, and thus all partitions/tasks will unfortunately need to be aborted as well. 

The API is a bit misleading since it implies you can choose which offsets to send to the transaction, but I recall @mjsax mentioning that the transaction is applied across all partitions at a time

Reply:
Correct. You need to distinguish between the read and write path here: the producer has no idea from what input partitions we consume, and thus it can only offer a generic `addOffsetsToTransaction` API and it's the user's (ie, our) responsibility to provide the correct offsets. What we write into the output topic has nothing to do with what we consumed from the input though: Even if you omit offsets for some input partitions, you still would have the corresponding pending writes on the output topic, so just omitting the offsets from the map does not help you to abort the pending writes for those partitions. What writes are committed has nothing to do with what offsets you pass to the producer.